cpp/src/strings/utilities.cu

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+
+#include <cstring>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <utilities/error_utils.hpp>
+#include "./utilities.hpp"
+
+#include <rmm/rmm.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/transform_scan.h>
+#include <thrust/transform_reduce.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+// Used to build a temporary string_view object from a single host string.
+std::unique_ptr<string_view, std::function<void(string_view*)>>
+    string_from_host( const char* str, cudaStream_t stream )
+{
+    if( !str )
+        return nullptr;
+    size_type length = (size_type)std::strlen(str);
+
+    char* d_str;
+    RMM_TRY(RMM_ALLOC( &d_str, length, stream ));
+    CUDA_TRY(cudaMemcpyAsync( d_str, str, length,
+                              cudaMemcpyHostToDevice, stream ));
+    CUDA_TRY(cudaStreamSynchronize(stream));
+
+    auto deleter = [](string_view* sv) { RMM_FREE(const_cast<char*>(sv->data()),0); };
+    return std::unique_ptr<string_view,
+        decltype(deleter)>{ new string_view(d_str,length), deleter};
+}
+
+// build an array of string_view objects from a strings column
+rmm::device_vector<string_view> create_string_array_from_column(
+    cudf::strings_column_view strings,
+    cudaStream_t stream )
+{
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_column = *strings_column;
+
+    auto count = strings.size();
+    rmm::device_vector<string_view> strings_array(count);
+    string_view* d_strings = strings_array.data().get();
+    thrust::for_each_n( execpol->on(stream),
+        thrust::make_counting_iterator<size_type>(0), count,
+        [d_column, d_strings] __device__ (size_type idx) {
+            if( d_column.nullable() && d_column.is_null(idx) )
+                d_strings[idx] = string_view(nullptr,0);
+            else
+                d_strings[idx] = d_column.element<string_view>(idx);
+        });
+    return strings_array;
+}
+
+// build a strings offsets column from an array of string_views
+std::unique_ptr<cudf::column> offsets_from_string_array(
+    const rmm::device_vector<string_view>& strings,
+    cudaStream_t stream, rmm::mr::device_memory_resource* mr )
+{
+    size_type count = strings.size();
+    auto d_strings = strings.data().get();
+    auto execpol = rmm::exec_policy(stream);
+    // offsets elements is the number of strings + 1
+    auto offsets_column = make_numeric_column( data_type{INT32}, count+1,
+                                               mask_state::UNALLOCATED,
+                                               stream, mr );
+    auto offsets_view = offsets_column->mutable_view();
+    auto d_offsets = offsets_view.data<int32_t>();
+    // create new offsets array -- last entry includes the total size
+    thrust::transform_inclusive_scan( execpol->on(stream),
```

Comment:
The UnaryFunction op (and parameters) will be very different here for each strings operation. There is likely some reasonable refactoring here but I'd like to put that off for a future PR if that is ok.

Reply:
It is not a question of difficulty. It certainly looks that way with only the 3 functions in this PR. There will be other items to consider in upcoming PRs and I'd rather not write this twice.

Reply:
I was thinking you can just use transform iterators and avoid needing to pass in the UnaryOp explicitly.

@davidwendt here's something to get you started. I think we should do this now instead of adding more tech debt for later:

```
std::unique_ptr<cudf::column> offsets_from_string_vector(
    const rmm::device_vector<string_view>& strings, cudaStream_t stream,
    rmm::mr::device_memory_resource* mr) {
  auto transformer = [] __device__(string_view v) { return v.size_bytes(); };

  auto begin = thrust::make_transform_iterator(strings.begin(), transformer);

  return make_offsets(begin, begin + strings.size(), mr, stream);
}

template <typename Iterator>
std::unique_ptr<column> make_offsets(
    Iterator begin, Iterator end,
    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
    cudaStream_t stream = 0) {

   CUDF_EXPECTS(begin <= end, "Invalid iterator range");

   auto count = thrust::distance(begin, end);

  auto offsets_column = make_numeric_column(
      data_type{INT32}, count + 1, mask_state::UNALLOCATED, stream, mr);

  auto offsets_view = offsets_column.mutable_view();

  thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), begin, end,
                         ++(offsets_view.begin<int32_t>()));

  CUDA_TRY(cudaMemsetAsync(offsets_view.begin<int32_t>(), 0, sizeof(int32_t),
                           stream));

  return offsets_column;
}
```